app/models/sm_sms_campaign_webhook/campaign_engagement.rb

```diff
@@ -0,0 +1,217 @@
+          raise InvalidPayload,
+                "campaign id missing from payload #{payload.inspect}"
+        end
+      )
+    rescue ArgumentError
+      raise InvalidPayloadValue,
+            "campaign id has invalid integer value #{payload.inspect}"
+    end
+
+    # @return [String] Keyword of the engaged campaign
+    # @raise [InvalidPayload] when campaign keyword missing from payload
+    def campaign_keyword
+      @campaign_keyword ||= String(
+        campaign_hash.fetch("keyword") do
+          raise InvalidPayload,
+                "campaign keyword missing from payload #{payload.inspect}"
+        end.freeze
+      )
+    end
+
+    # @return [Integer] ID of the engaging phone
+    # @raise [InvalidPayload] when phone id missing from payload
+    # @raise [InvalidPayloadValue] when phone id not numeric
+    def phone_id
+      @phone_id ||= Integer(
+        phone_hash.fetch("id") do
+          raise InvalidPayload,
+                "phone id missing from payload #{payload.inspect}"
+        end
+      )
+    rescue ArgumentError
+      raise InvalidPayloadValue,
+            "phone id has invalid integer value #{payload.inspect}"
+    end
+
+    # @return [String] Phone number engaging the campaign
+    # @raise [InvalidPayload] when phone number missing from payload
+    def phone_number
+      @phone_number ||= String(
+        phone_hash.fetch("number") do
+          raise InvalidPayload,
+                "phone number missing from payload #{payload.inspect}"
+        end.freeze
+      )
+    end
+
+    # @return [Integer] ID of campaign engagement state record
+    # @raise [InvalidPayload] when phone_campaign_state id missing from payload
+    # @raise [InvalidPayloadValue] when phone_campaign_state id not numeric
+    def phone_campaign_state_id
+      @phone_campaign_state_id ||= Integer(
+        phone_campaign_state_hash.fetch("id") do
+          raise InvalidPayload,
+                "phone_campaign_state id missing from payload #{payload.inspect}"
+        end
+      )
+    rescue ArgumentError
+      raise InvalidPayloadValue,
+            "phone_campaign_state id has invalid integer value #{payload.inspect}"
+    end
+
+    # @return [TrueClass,FalseClass] Has campaign engagement completed?
+    # @raise [InvalidPayload] when phone_campaign_state completed missing from payload
+    # @raise [InvalidPayloadValue] when phone_campaign_state completed is not boolean
+    def phone_campaign_state_completed?
+      # Has the boolean value already been assigned?
+      if !@phone_campaign_state_completed.nil?
+        return @phone_campaign_state_completed
+      end
+
+      # Extract the value and memoize it.
+      @phone_campaign_state_completed = begin
+        completed = phone_campaign_state_hash
+          .fetch("completed") do
+            raise InvalidPayload,
+                  "phone_campaign_state completed missing from payload #{payload.inspect}"
+          end
+
+        # Is this a boolean value?
```

Comment:
I assume you validated our JSON coming over uses true/false vs "true" or "false"?

Reply:
Yup, I doubled checked the value in example payloads during SMS campaign stuff:

```ruby
Started POST "/api/sms_campaign/webhook" for 127.0.0.1 at 2019-07-15 12:34:03 -0500
Processing by SmsWebhookController#create as JSON
  Parameters: {"created_at"=>"2019-07-15T17:34:02Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"9c0241cb-0935-4524-818a-19cd606f8571", "sms_webhook"=>{"created_at"=>"2019-07-15T17:34:02Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"9c0241cb-0935-4524-818a-19cd606f8571"}}
SmsWebhookController - payload: {"created_at"=>"2019-07-15T17:34:02Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"9c0241cb-0935-4524-818a-19cd606f8571"}
Completed 204 No Content in 1ms (ActiveRecord: 0.0ms | Allocations: 241)


Started POST "/api/sms_campaign/webhook" for 127.0.0.1 at 2019-07-15 12:34:15 -0500
Processing by SmsWebhookController#create as JSON
  Parameters: {"created_at"=>"2019-07-15T17:34:15Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"872fabde-9392-4a15-8494-e8f7f8cf355f", "sms_webhook"=>{"created_at"=>"2019-07-15T17:34:15Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"872fabde-9392-4a15-8494-e8f7f8cf355f"}}
SmsWebhookController - payload: {"created_at"=>"2019-07-15T17:34:15Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}}, "completed"=>false, "completed_at"=>nil, "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"872fabde-9392-4a15-8494-e8f7f8cf355f"}
Completed 204 No Content in 1ms (ActiveRecord: 0.0ms | Allocations: 253)


Started POST "/api/sms_campaign/webhook" for 127.0.0.1 at 2019-07-15 12:34:25 -0500
Processing by SmsWebhookController#create as JSON
  Parameters: {"created_at"=>"2019-07-15T17:34:25Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}, "vote"=>{"collected_at"=>"2019-07-15T17:34:25.134600Z", "value"=>2}}, "completed"=>true, "completed_at"=>"2019-07-15T17:34:25Z", "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"ab19992c-8913-4d4d-8317-e89911f2bcf7", "sms_webhook"=>{"created_at"=>"2019-07-15T17:34:25Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}, "vote"=>{"collected_at"=>"2019-07-15T17:34:25.134600Z", "value"=>2}}, "completed"=>true, "completed_at"=>"2019-07-15T17:34:25Z", "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"ab19992c-8913-4d4d-8317-e89911f2bcf7"}}
SmsWebhookController - payload: {"created_at"=>"2019-07-15T17:34:25Z", "data"=>{"campaign"=>{"id"=>10, "keyword"=>"WEBHOOKTEST1"}, "phone"=>{"id"=>1, "number"=>"6155890694"}, "phone_campaign_state"=>{"answers"=>{"email"=>{"collected_at"=>"2019-07-15T17:34:15.203635Z", "value"=>"cameron@southernmade.com"}, "vote"=>{"collected_at"=>"2019-07-15T17:34:25.134600Z", "value"=>2}}, "completed"=>true, "completed_at"=>"2019-07-15T17:34:25Z", "id"=>9}}, "object"=>"event", "type"=>"campaign.engagement", "uuid"=>"ab19992c-8913-4d4d-8317-e89911f2bcf7"}
Completed 204 No Content in 1ms (ActiveRecord: 0.0ms | Allocations: 265)
```